core_java/twitter/src/main/java/ca/jrvs/apps/twitter/service/TwitterService.java

```diff
@@ -0,0 +1,166 @@
+package ca.jrvs.apps.twitter.service;
+
+import ca.jrvs.apps.twitter.dao.TwitterDao;
+import ca.jrvs.apps.twitter.model.GeoLoc;
+import ca.jrvs.apps.twitter.model.Tweet;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.text.Normalizer;
+import java.text.Normalizer.Form;
+import java.util.ArrayList;
+import java.util.List;
+import org.slf4j.LoggerFactory;
+
+/**
+ * TwitterService contains validation and some additional processing of Tweets being made through
+ * TwitterCLI.
+ */
+public class TwitterService implements Service {
+
+  private TwitterDao dao;
+
+  //Default Constructor
+  public TwitterService() {
+    dao = new TwitterDao();
+  }
+
+  //Alternative constructor for providing a custom TwitterDao
+  public TwitterService(TwitterDao dao) {
+    this.dao = dao;
+  }
+
+  /**
+   * Validate and post a user input Tweet.
+   *
+   * @param tweet tweet to be created
+   * @return created tweet
+   * @throws IllegalArgumentException if text exceed max number of allowed characters or lat/long
+   *                                  out of range
+   */
+  @Override
+  public Tweet postTweet(Tweet tweet) throws IllegalArgumentException {
+    String tweetText = tweet.getText();
+    GeoLoc coords = tweet.getLocation();
+    float latitude;
+    float longitude;
+    boolean validCoords = false;
+    boolean hasCoords = false;
+    boolean validText = false;
+
+    validText = validateText(tweetText);
+    if (coords != null) {
+      longitude = coords.getCoordinates()[0];
+      latitude = coords.getCoordinates()[1];
+      hasCoords = true;
+      validCoords = validateCoordinates(latitude, longitude);
+    }
+
+    if (validText) {
+      if (hasCoords && validCoords) {
+        return dao.create(tweet);
+      } else if (!hasCoords) {
+        return dao.create(tweet);
+      }
+    }
+    return null;
+  }
+
+  private boolean validateText(String tweetText) {
+    // Lazy length check. Twitter shortens URLs so actual tweet length may be shorter than this.
+    // Due to UTF8 encoding quirks, Normalizing the tweet text may change its length, Twitter uses
+    // NFC normalization before checking tweet length, so we do that too.
+    tweetText = Normalizer.normalize(tweetText, Form.NFC);
+    if (tweetText.length() > 280) {
```

Comment:
On the handout it says tweets should be 140 characters or less

Reply:
Twitter updated that limit to 280 chars a couple years ago